engine/src/flutter/impeller/renderer/backend/vulkan/render_pass_vk.cc

```diff
@@ -146,40 +145,37 @@ RenderPassVK::RenderPassVK(const std::shared_ptr<const Context>& context,
     return true;
   });
 
-  SharedHandleVK<vk::RenderPass> recycled_render_pass;
-  SharedHandleVK<vk::Framebuffer> recycled_framebuffer;
-  if (resolve_image_vk_) {
-    recycled_render_pass =
-        TextureVK::Cast(*resolve_image_vk_).GetCachedRenderPass();
-    recycled_framebuffer =
-        TextureVK::Cast(*resolve_image_vk_).GetCachedFramebuffer();
-  } else {
-    recycled_render_pass =
-        TextureVK::Cast(*color_image_vk_).GetCachedRenderPass();
-    recycled_framebuffer =
-        TextureVK::Cast(*color_image_vk_).GetCachedFramebuffer();
-  }
-
-  const auto& target_size = render_target_.GetRenderTargetSize();
-
+  FramebufferAndRenderPass frame_data;
   bool is_swapchain = false;
+  SampleCount sample_count =
+      color_image_vk_->GetTextureDescriptor().sample_count;
   if (resolve_image_vk_) {
+    frame_data = TextureVK::Cast(*resolve_image_vk_).GetCachedFrameData();
     is_swapchain = TextureVK::Cast(*resolve_image_vk_).IsSwapchainImage();
   } else {
+    frame_data = TextureVK::Cast(*color_image_vk_).GetCachedFrameData();
     is_swapchain = TextureVK::Cast(*color_image_vk_).IsSwapchainImage();
   }
+  // If the sample count changed from the last cache usage, clear the
+  // cached framebuffer and render pass.
+  if (sample_count != frame_data.sample_count) {
+    frame_data.framebuffer = nullptr;
+    frame_data.render_pass = nullptr;
+  }
+
```

Comment:
Instead of clearing out the cached data if it doesn't have the right sample counts, should we instead be caching based on the sample count?  This will cause thrashing in the cache in playgrounds where something is cached then thrown away directly, no?  This isn't a big deal for playgrounds, my concern is that we'd accidentally introduce this behavior in production and it would take effort to figure out.

Reply:
I think its unlikely to be a problem in a real application as the render_target_cache actually separates MSAA and non-MSAA allocations.

Reply:
By effectively not caching in playgrounds but caching in production we are opening ourselves up to bugs that can only be reproduced in production, no?

Reply:
this is only going to happen with the imgui overlay active though.

Reply:
In playgrounds we do the imgui pass regardless if there is some imgui overlay active though.  The test I was using was `CanDrawCapsAndJoins` which doesn't have any visual change from imgui.

The golden tests don't have the imgui pass though.

This feels fragile to me.  What argument is there against adding `TextureSourceVK::GetCachedFrameData(size_t sample_count)`?

Alternatively, I think instead of nulling out the cached value when it doesn't match our needs we can just put it back.  That way we wouldn't be thrashing.  We'd still be caching when it's possible and creating a new one when it isn't.  That makes the worst case scenario less worse.

Reply:
We don't use it for the goldens though, and AFAIK that is what is actually running on CI?